File: src/net/eduard/api/lib/bungee/BukkitMessageListener.java
package net.eduard.api.lib.bungee;

import org.bukkit.entity.Player;
import org.bukkit.plugin.messaging.PluginMessageListener;


import java.io.ByteArrayInputStream;
import java.io.DataInputStream;
import java.io.IOException;
import java.io.ObjectInputStream;

public class BukkitMessageListener implements PluginMessageListener {

    public BukkitMessageListener(BukkitController bukkitController) {
        setController(bukkitController);
    }

    private BukkitController controller;

    @Override
    public void onPluginMessageReceived(String channel, Player player, byte[] message) {
        if (channel.equals(BungeeAPI.getChannel())) {
            ByteArrayInputStream arrayIn = new ByteArrayInputStream(message);

            DataInputStream data = new DataInputStream(arrayIn);
            try {
                ObjectInputStream objStream = new ObjectInputStream(arrayIn);
                String server = objStream.readUTF();
                String tag = objStream.readUTF();
                if (tag.equals("server-update")){

                    ServerSpigot spigot = (ServerSpigot) objStream.readObject();
                    BungeeAPI.getServers().
                            put(spigot.getName().toLowerCase(),spigot);


                }else {
                    String line = data.readUTF();
                    controller.receiveMessage(server, tag, line);
                }
            } catch (IOException | ClassNotFoundException e) {
                e.printStackTrace();
            }

        }
    }

    public BukkitController getController() {
        return controller;
    }

    public void setController(BukkitController controller) {
        this.controller = controller;
    }

}

File: src/main/java/net/eduard/api/lib/bungee/BukkitMessageListener.kt
package net.eduard.api.lib.bungee

import com.google.common.util.concurrent.Futures
import java.io.ByteArrayInputStream
import java.io.DataInputStream
import java.io.ObjectInputStream
import org.bukkit.entity.Player
import org.bukkit.plugin.messaging.PluginMessageListener
import java.io.IOException
import java.lang.ClassNotFoundException
import java.util.concurrent.CompletableFuture

class BukkitMessageListener(var bukkitController: BukkitController) : PluginMessageListener {
    fun readMessage(arrayIn: ByteArrayInputStream, data: DataInputStream) {
        try {
            val objStream = ObjectInputStream(arrayIn)
            val server = objStream.readUTF()
            val tag = objStream.readUTF()
            if (tag == "server-update") {
                val spigot = objStream.readObject() as ServerSpigot
                BungeeAPI.servers[spigot.name.toLowerCase()] = spigot
            } else {
                val line = data.readUTF()
                bukkitController.receiveMessage(server, tag, line)
            }
        } catch (e: IOException) {
            e.printStackTrace()
        } catch (e: ClassNotFoundException) {
            e.printStackTrace()
        } catch (e: ConcurrentModificationException) {
            e.printStackTrace()
        }
    }

    override fun onPluginMessageReceived(channel: String, player: Player, message: ByteArray) {
        if (channel == BungeeAPI.channel) {
            val arrayIn = ByteArrayInputStream(message)
            val data = DataInputStream(arrayIn)
            CompletableFuture.runAsync {
                readMessage(arrayIn, data)
            }
        }
    }

}